src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs

```diff
@@ -194,38 +194,57 @@ internal static void SetLogRecordSeverityFields(ref LogRecordData logRecordData,
             iLoggerData.State = state;
             return null;
         }
+#if NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER
+        else if (RuntimeFeature.IsDynamicCodeSupported)
+        {
+            return ParseCustomState(logRecord, state);
+        }
         else
         {
-            try
-            {
-                PropertyDescriptorCollection itemProperties = TypeDescriptor.GetProperties(state);
+            OpenTelemetrySdkEventSource.Log.LoggerProcessStateSkipped<TState>();
+            return Array.Empty<KeyValuePair<string, object?>>();
+        }
+#else
+        else
+        {
+            return ParseCustomState(logRecord, state);
+        }
+#endif
+    }
 
-                var attributeStorage = logRecord.AttributeStorage ??= new List<KeyValuePair<string, object?>>(itemProperties.Count);
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "ParseCustomState will only be called when RuntimeFeature.IsDynamicCodeSupported is true.")]
```

Comment:
This isn't correct. `RuntimeFeature.IsDynamicCodeSupported` is for "AOT".  But `IL2026` is for "trimming". They are 2 different ways to publish your app. Yes, when you "AOT" your app, it is also "trimmed".  But you can just "trim" your app without AOT'ing it. That is what `IL2026` is warning you about. Even though you guarantee you won't call this method when AOT'd, it can still be called in a trimmed app and you will see runtime behavior differences (i.e. `TypeDescriptor.GetProperties` will return a different set of properties).

Reply:
In fact - if the conditions are correct you should not need any suppressions anywhere. The fact that you had to add a suppression means that there's a problem here. And it's exactly what Eric pointed out above.

I still think that the only solution will be a feature switch - we can tie that to trimming and it would end up acting the way you want.

The other observation - the code here is actually AOT compatible even if does the reflection - from pure AOT perspective there are no issues. So if we ever had an AOT compiler which doesn't trim, this would work just fine with that. So tying this to IsDynamicCodeSupported is technically wrong - you're relying on an assumption the AOT implies trimming. That currently is the case, but it may not be in the future (and it already sort of isn't for targets like Mono-Android/iOS in some cases).

Reply:
Shall we park this PR for now until we get a better clarity on the feature switch solution?

Reply:
I'll prototype a feature switch for this, but the overall shape of the change looks good. We just can't use `IsDynamicCodeSupported` it will be replaced with the new feature switch.

Reply:
@vitek-karas and @eerhardt:
Due to the nature of trimmer, when there is no object holding reference to the property, it will be trimmed out - we assume  there will be no behavior difference from the SDK side and is okay with this indeterministic behavior at runtime for the use case of only trimmer enabled but not AOT'ed app. Based on this assumption, there will be no harm in using [UnconditionalSuppressMessage] here. What do you think?

Just out of curiousity, is it technically-viable to introduce a property for trimmer, as what we have today `.IsDynamicCodeSupported` for AOT?

Reply:
In general our trimming and AOT story tries to be predictable. It's kind of the pri 0. in the past we've had trimming experiences which were not fully predictable and it was (and still is) causing lot of confusion.
That's why we have annotations and static analysis and warnings - the goal is to make it super predictable: "If I don't get warnings, the app will work".

Unfortunately we can't have the 100% predictability via static analysis in all cases (or rather: it's prohibitively expensive to make it so). So in some very specific and carefully considered cases we try to achieve the next best thing: predictable runtime behavior. That means that if I turn on trimming/AOT on my app, the app may slightly change behavior, but it will have consistent behavior across all configurations - mainly the debug build (`dotnet build`, the `F5` in VS and so on) will behave the same as published version of the app.

What you're proposing is neither, it's the unpredictable behavior we're trying to avoid - the app works in debug build, I don't get any warnings and when I publish it, the app doesn't work the same.

Reply:
> Just out of curiousity, is it technically-viable to introduce a property for trimmer, as what we have today .IsDynamicCodeSupported for AOT?

It is definitely technically viable; we've tried to avoid adding one because it would make it easy to write code which changes behavior based on trimming on/off and we really want people to try hard to achieve the "statically analyzable" part of the comment above.

AOT compatibility issues are much less common than trimming compatibility and also the property already existed before for other reasons (it has performance implications, unlike trimming which really doesn't have other implications).

That's not to say that we won't every have it, but currently we don't. And even if we add it in .NET 8, it's not going to be usable right now.

The "Workaround" and our currently preferred option (if there's no way to make the code statically correct) is to use feature switches. They achieve basically the same thing, but they can be domain specific - and sometimes they're useful outside of trimming context.

In this context, normally this would be solvable by marking the setting which enables this as `RequiresUnreferencedCode`. Unfortunately that setting is part of "options" and so it's set automatically through the configuration system - and thus it's effectively impossible to warn about it statically (since the values come at runtime). That's why something like a feature switch would make the most sense here - at least to me. It has an unfortunate side-effect that it would require a small MSBuild integration for the SDK NuGet package (to correctly setup the default value of that feature switch).

Reply:
An example of something that breaks here in trimming is when the app only sets a property, but never gets it in code. The above code breaks because the getter is trimmed. And so the the Property value can't be retrieved.

Reply:
Drop support the branch with `TypeDescriptor.GetProperties(State)` which was introduced in this PR: https://github.com/open-telemetry/opentelemetry-dotnet/pull/4334/files#diff-78727a20b418c7cda4068e10bd56a5fe678014f76dc8ac305a04db446f3d7165,
given 99.9% of TState will implement IReadOnlyList - the source generator used in [Microsoft.Extensions.Logging ILogger.Log<TState> ](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.log?view=dotnet-plat-ext-7.0) guarantees TState implements IReadOnlyList.

Reply:
Let me try to simplify/clarify our discussion from the SIG a bit. What this code is doing is capturing properties of objects which are logged. If those objects are being trimmed, that's fine. We only want to log the stuff that is being used in the application and not trimmed anyway. So our thinking here is that if `TypeDescriptor.GetProperties` is able to return the valid non-trimmed stuff, that works for us we can just suppress the warning that it might not return _everything_. Properties on the logged objects which aren't being used and have `null` values we drop anyway.

Does that make sense? Valid thinking or no?

Reply:
The thing to be aware of is that you can't guarantee:

> We only want to log the stuff that is being used in the application and not trimmed anyway.

There are 2 cases where this isn't guaranteed:

1. My example [above](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4560#discussion_r1230074282). The trimmer only sees the property being set, and trims the `getter`. In this case the property is used by the application (it is set), but the value won't be logged.
2. In NativeAOT, the AOT compiler is free to trim any metadata it doesn't see as being Reflected against. In this case, the property can be both `set` and `get` in the application, but the AOT compiler trimmed away the property metadata because it didn't think it was needed. No one said to preserve the properties of this Type. Here the stuff is definitely being used in the app (the property was both get and set), however the Property metadata isn't there and this code doesn't log it.

So it depends on how OK you are with the above 2 situations. The app behavior will change once it has been AOT'd, and our goal is that if an app developer doesn't get any warnings, they are guaranteed the app will work the same after trimming/AOT as it did before trimming/AOT.

Reply:
Discussed in 6/20 SIG meeting, maintainers aligned to drop support for ParseStateValues if the state neither implements IReadOnlyList nor IEnumberable based on the above discussions in this thread.